BoxContentSDK/BoxContentSDK/Views/BOXUserAvatarImageView.m

```diff
@@ -0,0 +1,267 @@
+//
+//  BOXUserAvatarImageView.m
+//  BoxContentSDK
+//
+//  Created by Rico Yao on 9/27/17.
+//  Copyright © 2017 Box. All rights reserved.
+//
+
+#import "BOXUserAvatarImageView.h"
+#import "BOXUserAvatarRequest.h"
+#import "BOXContentClient+User.h"
+#import "BOXContentCacheClientProtocol.h"
+
+typedef enum {
+    BOXCircularAvatarBackgroundShadeSoftMagenta = 0,
+    BOXCircularAvatarBackgroundShadeSoftCyan,
+    BOXCircularAvatarBackgroundShadeLightRed,
+    BOXCircularAvatarBackgroundShadeModerateGreen,
+    BOXCircularAvatarBackgroundShadeVividViolet,
+    BOXCircularAvatarBackgroundShadeLightOrange,
+    BOXCircularAvatarBackgroundShadeBrightRed,
+    BOXCircularAvatarBackgroundShadeLimeGreen,
+    BOXCircularAvatarBackgroundShadeSoftPink,
+    BOXCircularAvatarBackgroundShadeModerateBlue,
+    BOXCircularAvatarBackgroundShadeDarkOrange,
+    BOXCircularAvatarBackgroundShadeDesaturatedBlue,
+    BOXCircularAvatarBackgroundShadeModerateLimeGreen,
+    BOXCircularAvatarBackgroundShadeBrightYellow,
+    BOXCircularAvatarBackgroundShadeSoftBlue,
+    BOXCircularAvatarBackgroundShadeDarkViolet,
+    BOXCircularAvatarBackgroundShadeBrightOrange,
+    BOXCircularAvatarBackgroundShadeVividOrange,
+    BOXCircularAvatarBackgroundShadeCount,
+} BOXCircularAvatarBackgroundShade;
+
+@interface BOXUserAvatarImageView()
+@property (nonatomic, readwrite, strong) BOXUserAvatarRequest *request;
+@end
+
+@implementation BOXUserAvatarImageView
+
+- (instancetype)initWithFrame:(CGRect)frame
+                         user:(BOXUserMini *)user
+                       client:(BOXContentClient *)client
+{
+    if (self = [super initWithFrame:frame]) {
+        _user = user;
+        _client = client;
+        
+        self.contentMode = UIViewContentModeScaleAspectFill;
+        self.layer.masksToBounds = YES;
+
+        if (user) {
+            [self render];
+        }
+    }
+    return self;
+}
+
+- (instancetype)initWithFrame:(CGRect)frame
+                       client:(BOXContentClient *)client
+{
+    self = [self initWithFrame:frame user:nil client:client];
+    return self;
+}
+
+- (void)layoutSubviews
+{
+    [super layoutSubviews];
+    self.layer.cornerRadius = self.bounds.size.width / 2;
+}
+
+- (void)setUser:(BOXUserMini *)user
+{
+    NSString *oldUserID = _user.modelID;
+    _user = user;
+    if (![oldUserID isEqualToString:user.modelID]) {
+        [self render];
+    }
+}
+
+#pragma mark - Private helpers
+
+- (void)render
+{
+    [self.request cancel];
+    self.request = [self.client userAvatarRequestWithID:self.user.modelID type:BOXAvatarTypeLarge];
```

Comment:
1. do you want to check for non null user and client objects and skip the flow if they are?
2. what happens if multiple instances of this class are requesting the user avatar for the same user? will it fire off multiple or have some way to check for a pending request and block the rest?

Reply:
1. Good point on nil-user check. I'll add that. The nil-client check feels a bit unnecessary and it's not something we check for in the rest of the SDK. For most users the client is effectively a singleton so it's never really nil.

2. The requests go into an operation queue where we've established reasonable concurrent execution limits. So it should handle many requests getting fired off pretty well.